src/qibolab/unrolling.py

```diff
@@ -0,0 +1,79 @@
+"""Utilities for sequence unrolling.
+
+May be reused by different instruments.
+"""
+
+from dataclasses import asdict, dataclass, field, fields
+from functools import total_ordering
+
+from .pulses import PulseSequence
+
+
+def _waveform(sequence: PulseSequence):
+    # TODO: deduplicate pulses (Not yet as drivers may not support it yet)
+    # TODO: count Rectangular and delays separately (Zurich Instruments supports this)
+    # TODO: Any constant part of a pulse should be counted only once (Zurich Instruments supports this)
+    # TODO: check if readout duration is faithful for the readout pulse (I would only check the control pulses)
+    # TODO: Handle multiple qubits or do all devices have the same memory for each channel ?
+    return sequence.duration - sequence.ro_pulses.duration
+
+
+def _readout(sequence: PulseSequence):
+    # TODO: Do we count 1 readout per pulse or 1 readout per multiplexed readout ?
+    return len(sequence.ro_pulses)
+
+
+def _instructions(sequence: PulseSequence):
+    return len(sequence)
+
+
+@total_ordering
+@dataclass(frozen=True, eq=True)
+class Bounds:
+    """Instument memory limitations proxies."""
+
+    waveforms: int = field(metadata={"count": _waveform})
+    """Waveforms estimated size."""
```

Comment:
is this number of samples?

Reply:
Nope, you could check the `TODO` above, and the discussion in:
https://github.com/qiboteam/qibolab/pull/817/files/78b5103af3eb20fd9d7f29485176d55a19857837#r1502587002

It is the estimate of the waveforms size memory, whatever it means. In practice, I'd say that it should be the number of samples for deduplicated non-rect (thus non-delay) pulses.
However, we don't know exactly, and there could even be more complications around. So, I'd keep the original definition: *estimate of the waveform memory*, whatever it means.

Reply:
For now I just get the sequence, remove the readout and check how  long it lasts. It is not the best estimate but for this PR gets something that it is better than nothing for all the drivers. We should later work on improving the bounds.